### DISCUSSION TITLE
I can't make post request to my .json files in Deployment

### DISCUSSION THREAD

**Discussion by Necoo33** [vercel/next.js#46021] [Help] [answered]
Hello, i created a website for me and build api's for that, which uses .json databases. i'm deployed it and in vercel production, i could do get requests succesfully but when i try to make post requests i got the "Error: EROFS: read-only file system" error. How can i fix that?

**Comment by icyJoseph** [comment, marked as answer]
There's a couple of problems here.

First of all, there's a limit to how much memory your deployments can consume, in free/hobby tier anyway, and writing to a JSON file, at runtime might increase that beyond the limit.

Deployments in Vercel are serverless, and are run as tasks, which are independent units of works. 

- https://github.com/orgs/vercel/discussions/241#discussioncomment-363002

Also API route instances are destroyed after a given timeout.

When you have request A coming to your endpoint, an instance would be created, including the JSON file you included as Database, if you were allowed to write then, when the instance is killed the changes you made would be destroyed with it too. 

Also, two calls coming to the server might be served by different instances, so the JSON data you are modifying won't be the same, anyway.

To prevent these pitfalls, and others (security related, concurrency, etc.), the task running environment disallows writing to disk. 

You might still write to `/tmp` for various other kinds of data, like receiving uploads and such, but the serverless environment is not designed to hold data itself.

- https://vercel.com/guides/how-can-i-use-files-in-serverless-functions#using-temporary-storage
- https://github.com/vercel/community/discussions/314?sort=new

If you managed your own Next.js deploy, then you could make it a regular server, by just calling `next start`.

**Reply by Necoo33** [reply]
Thanks for the informating answer. I have some counter questions which i curious about:

1 - if i put my files into /tmp folder, how do i write them, basically using path.join(), process.cwd() and writeFileSync() syntax?

2 - if i follow /tmp way will my post added datas delete a while later or has it only a limit or not?

3 - if i use services alike cloud firestore or mongoDB could i write that databases?

4 - if i deploy my website with another service, like firebase host, will i face like that problems or not or it depends to their hosting way?

5 - how to make my deploy regular server?

**Reply by icyJoseph** [reply]
> 1 - if i put my files into /tmp folder, how do i write them, basically using path.join(), process.cwd() and writeFileSync() syntax?
> 2 - if i follow /tmp way will my post added datas delete a while later or has it only a limit or not?

As far as I know, the tmp folder will disappear with the function instance... but I am not 100% sure, I wouldn't rely on this anyway

> 3 - if i use services alike cloud firestore or mongoDB could i write that databases?

Yeah because then your API route would post to their DB service, and then they'd hold that data. Your API route will then be shutdown, and new requests would read the data from the DB service instead.

> 4 - if i deploy my website with another service, like firebase host, will i face like that problems or not or it depends to their hosting way?
> 5 - how to make my deploy regular server?

Depends, some service providers favour serverless deployments because they are very cheap, compared to having a dedicated computer running a server waiting for requests. 

It might be that in the case of Firebase they use Cloud Functions, or some other technique, and then you'll just have the same error. And the solution would be the same, to use a DB service.

AWS Amplify uses serverless too when deploying Next.js app, again because it is very cost efficient to do so.

So you have to figure out if the hosting provider will use a Serverless deployment for your Next.js app.

To deploy your server as a "regular" Node.js application, you need a service provider that can host Node.js applications, that's it. Clone your project into it, build it and run it.

If you buy an EC2 in AWS, git clone your project, build it and run it, then you can do what you want, but that's also buying yourself into more problems and costs.
